### ISSUE TITLE
Proposed changes to search UI

### ISSUE THREAD

**Issue by jart** [tensorflow/tensorboard#489]
Currently on each dashboard:

- An "all tags" section is visible, i.e. "Tags matching: /.*/"
- Tag sections are collapsed
- Entering a search causes the "all tags" section to drill down
- Clearing the search box causes the "all tags" section to disappear

We've received feedback from multiple individuals that the first thing they do in their workflows is collapse the first section and open the other ones.

It might be better to change the UI to the following:

1. Remove "all tags" section
2. Expand all sections by default (should be possible, thanks to pagination)
3. Make search bar invisible by default (free up vertical whitespace: content is king)
4. Add magnifying glass icon (left to refresh button) that toggles visibility of search bar
5. When initially clicked, search bar initializes to empty (defined as `.*`) and "all tags" section appears
6. Typing search term causes the "Tags matching" section to update in real time, as it does now
7. *Maybe* put "x" button to right of search bar that does same thing as clicking search button in top right corner a second time (de-toggling)
8. Ensure that all these search states persist to the fragment section of the URL bar

Thoughts @wchargin @chihuahua @dandelionmane?

**Comment by jart** [comment]
Alternatively, we could have the search apply to all sections. Each section title could say "Tag Name <grey>matching /foo/</grey>" as soon as "foo" is entered in the search box. Then any sections with zero matching items would disappear. If all sections disappear, then a piece of text shows up that says "No results."

**Comment by jart** [comment]
Another thing we could do with the search bar, to free up space for content, is simply put it at the top of the side bar.

**Comment by wchargin** [comment]
Sounds good to me.

Comment#1's approach doesn't let you search across tag groups, which we might want to keep.

No opinion as to whether side bar or main pane is better.

One note: persisting everything to the URL bar also suggests sharing the search query across all plugins. It's not clear that this is the right option, as each tag name appears only within one plugin (well, except for distributions/histograms). It would also be possible to persist `scalarSearchQuery=`, etc. to the URL bar individually.

**Comment by gtoderici** [comment]
With regards to making ".*" the default: it's counter-intuitive to default to something that most casual users won't know how to interpret. In order to know that ".*" is "match everything" the user would know that this is a regular expression, which means that a large number of users will be confused by it. Defaulting to "" (nothing) and having that display all tags would be a less confusing solution.

Other than this, having all sections expanded by default would be good. 

The search bar being placed on the side bar with a clear button for "searching tags" will also be an improvement since it will free up space.

**Comment by wchargin** [comment]
@gtoderici: We explicitly write "(all tags)" next to the query when the query is `.*` for exactly this reason. We also write the query as `/.*/` to signal that it is a regular expression.

**Comment by gtoderici** [comment]
I agree that this will signal that this is a regex to people who know what a regex is. However, many people (undergrad students, for example), may have never encountered this, and will simply be confused.

**Comment by jart** [comment]
This could be solved by the following:

- Use `placeholder="regex"` on search box. (Magnifying glass communicates it is for search.)
- Say "All tags" in section header when search box is empty.
- Say "Tags matching: /FOO/" when search box is non-empty.

That way it'll make sense even if the user doesn't know what regex is. He'll also have a hint about what to google.
